keps/sig-api-machinery/20190507-kube-controller-manager-config.md

```diff
@@ -0,0 +1,478 @@
+  HorizontalPodAutoscalerDownscaleForbiddenWindow: 0s
+  HorizontalPodAutoscalerDownscaleStabilizationWindow: 5m0s
+  HorizontalPodAutoscalerInitialReadinessDelay: 30s
+  HorizontalPodAutoscalerSyncPeriod: 15s
+  HorizontalPodAutoscalerTolerance: 0.1
+  HorizontalPodAutoscalerUpscaleForbiddenWindow: 0s
+  HorizontalPodAutoscalerUseRESTClients: true
+JobController:
+  ConcurrentJobSyncs: 4
+KubeCloudShared:
+  AllocateNodeCIDRs: false
+  AllowUntaggedCloud: false
+  CIDRAllocatorType: RangeAllocator
+  CloudProvider:
+    CloudConfigFile: ''
+    Name: ''
+  ClusterCIDR: ''
+  ClusterName: kubernetes
+  ConfigureCloudRoutes: true
+  ExternalCloudVolumePlugin: ''
+  NodeMonitorPeriod: 5s
+  NodeSyncPeriod: 0s
+  RouteReconciliationPeriod: 10s
+  UseServiceAccountCredentials: true
+NamespaceController:
+  ConcurrentNamespaceSyncs: 10
+  NamespaceSyncPeriod: 5m0s
+NodeIPAMController:
+  NodeCIDRMaskSize: 24
+  ServiceCIDR: ''
+NodeLifecycleController:
+  EnableTaintManager: true
+  LargeClusterSizeThreshold: 50
+  NodeEvictionRate: 0.1
+  NodeMonitorGracePeriod: 40s
+  NodeStartupGracePeriod: 1m0s
+  PodEvictionTimeout: 5m0s
+  SecondaryNodeEvictionRate: 0.01
+  UnhealthyZoneThreshold: 0.55
+PersistentVolumeBinderController:
+  PVClaimBinderSyncPeriod: 15s
+  VolumeConfiguration:
+    EnableDynamicProvisioning: true
+    EnableHostPathProvisioning: false
+    FlexVolumePluginDir: "/usr/libexec/kubernetes/kubelet-plugins/volume/exec/"
+    PersistentVolumeRecyclerConfiguration:
+      IncrementTimeoutHostPath: 30
+      IncrementTimeoutNFS: 30
+      MaximumRetry: 3
+      MinimumTimeoutHostPath: 60
+      MinimumTimeoutNFS: 300
+      PodTemplateFilePathHostPath: ''
+      PodTemplateFilePathNFS: ''
+PodGCController:
+  TerminatedPodGCThreshold: 12500
+ReplicaSetController:
+  ConcurrentRSSyncs: 5
+ReplicationController:
+  ConcurrentRCSyncs: 5
+ResourceQuotaController:
+  ConcurrentResourceQuotaSyncs: 5
+  ResourceQuotaSyncPeriod: 5m0s
+SAController:
+  ConcurrentSATokenSyncs: 5
+  RootCAFile: "/var/run/kubernetes/server-ca.crt"
+  ServiceAccountKeyFile: "/tmp/kube-serviceaccount.key"
+ServiceController:
+  ConcurrentServiceSyncs: 1
+TTLAfterFinishedController:
+  ConcurrentTTLSyncs: 5
+```
+
+Besides, we construct a configMap used to hold the component config, the struct like
+```
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  namespace: default
+  name: deployment-kube-controller-manager-config
+data:
+  config.yaml: |-
+```
+
+we can use the context of KubeControllerManagerConfiguration to fill the `config.yaml` field, so we can use it in cluster
+as a resource.
+
+### How is it versioned
+
+Currently, the controller manager's version is `Alpha`, with `register` and `scheme` function.
+
+## Design Details
+
+### Struct for flags
+
+The relative struct `KubeControllerManagerOptions` and `Config` file located in `cmd/kube-controller-manager/app` directory,
+the struct define as below:
+
+```
+type KubeControllerManagerOptions struct {
```

Comment:
dumping everything in a single struct makes it harder to run controllers in different contexts (e.g., from different binaries, as in the cloud controller extraction effort) .